Chapter 18: Texture Sampling

AristurtleDev opened the pull request:
Live Preview https://aristurtle.net/tutorials/building_2d_games/18_texture_sampling/

AristurtleDev committed:
wip

AristurtleDev committed:
Use fuel cell for anisotrpic comparison

AristurtleDev committed:
Initial draft

djjohnsongeek commented:
I noticed this message in [chapter 4:](https://aristurtle.net/tutorials/building_2d_games/04_creating_a_class_library/?tabs=vscode#conclusion)
![image](https://github.com/user-attachments/assets/36cb3857-d186-402b-a937-acb4a2a94c6f)
However it seems out of place because the tutorial so far has not implemented an FPS counter. That or I missed it.

[Chapter 7 ](https://aristurtle.net/tutorials/building_2d_games/07_optimizing_texture_rendering/#using-the-textureatlas-class) also refers to this fps counter:
![image](https://github.com/user-attachments/assets/66db1e23-aa11-44fb-a097-33472b7c459e)

djjohnsongeek commented:
[Chapter 5](https://aristurtle.net/tutorials/building_2d_games/05_content_pipeline/?tabs=vscode#loading-our-first-asset) Typos:
![image](https://github.com/user-attachments/assets/fdb5234e-272a-46fd-8a40-11a12d8ae7e8)

djjohnsongeek commented:
In chapter 7 it seems like there are some duplicate KeyboardState properties in the [KeyboardInfo class](https://aristurtle.net/tutorials/building_2d_games/11_input_management/#keyboardinfo-properties).
![image](https://github.com/user-attachments/assets/ca777f4a-f9ff-4dd9-98ce-a6530b8d96db)

djjohnsongeek commented:
After defining the [MouseInfo methods](https://aristurtle.net/tutorials/building_2d_games/11_input_management/#mouseinfo-methods), the explanation text references the KeyboardInfo methods instead of  the mouse methods.
![image](https://github.com/user-attachments/assets/f284db6d-ca9f-464b-92c7-8023f13d7ec4)

djjohnsongeek commented:
Chapter 11, in the [Updating the Game1 Class](https://aristurtle.net/tutorials/building_2d_games/11_input_management/#updating-the-game1-class) section the "key changes" portion after the code indicates that "the check for the gamepad back button or keyboard escape key being pressed was removed" from the `Game1.Update)` method. This is not the case.
![image](https://github.com/user-attachments/assets/569ea105-8ee9-4d62-8785-0744fce60c6d)
![image](https://github.com/user-attachments/assets/d5413c17-113a-4bbf-b568-7be2d934461b)

djjohnsongeek commented:
Chapter 12, [Adding Collision to Our Game](https://aristurtle.net/tutorials/building_2d_games/12_collision_detection/#adding-collision-to-our-game). The code provided sets `_batPosition` in the `Initialize()` method using `_slime`, but this results in an error because `_slime` is not defined until the `LoadContent` method.
![image](https://github.com/user-attachments/assets/512fb8e8-9c70-422b-ae36-460ad5da9aed)

djjohnsongeek commented:
In Chapter 12, [Conclusion](https://aristurtle.net/tutorials/building_2d_games/12_collision_detection/#conclusion), the content indicates that in the the next chapter "we'll start exploring audio". However the next chapter is about tile maps.
![image](https://github.com/user-attachments/assets/2a37f32b-eecb-48fe-ad95-b76eed00cb69)

AristurtleDev committed:
Remove FPS counter references

AristurtleDev committed:
Fix teh typos

AristurtleDev committed:
Remove duplicate properties in KeyboardInfo

AristurtleDev committed:
Fix mouse method description list

AristurtleDev committed:
Use the verbage 'the specified key' instead of 'a key'

AristurtleDev committed:
Remove the exit check in game1

AristurtleDev committed:
Update conclusion since retroactivly adding additional chapter

AristurtleDev committed:
Merge branch 'fix/errors' into 18-texture-sampling

AristurtleDev committed:
Merge branch '18-texture-sampling' of https://github.com/aristurtledev/docs.monogame.github.io into 18-texture-sampling

AristurtleDev commented:
Thanks @djjohnsongeek for the second eyes on some of this.  I've updated the chapters to (hopefully) resolve the issues if you wanna double check again.  Some of these, like the FPSCounter were originally part of the tutorial, but it's been an evolving tutorial as I'm getting close to the finish line and some things get retroactively removed or added, and artifacts get left by accident.


In regards to 

> Chapter 12, [Adding Collision to Our Game](https://aristurtle.net/tutorials/building_2d_games/12_collision_detection/#adding-collision-to-our-game). The code provided sets `_batPosition` in the `Initialize()` method using `_slime`, but this results in an error because `_slime` is not defined until the `LoadContent` method. ![image](https://private-user-images.githubusercontent.com/43294314/428701602-512fb8e8-9c70-422b-ae36-460ad5da9aed.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDM0NjI2OTksIm5iZiI6MTc0MzQ2MjM5OSwicGF0aCI6Ii80MzI5NDMxNC80Mjg3MDE2MDItNTEyZmI4ZTgtOWM3MC00MjJiLWFlMzYtNDYwYWQ1ZGE5YWVkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMzElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzMxVDIzMDYzOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTExZWZlMDc0MzMxN2E5ZjBmMGQ1NmQ3YWRkODI2OGM1YzUzZGQ5OTE5ODBiYTdiZDE3NmM3NTA1YTI5Yjc0ZjMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.BWlHR0Aj9carK506pan53UIarCblFf1z_FbVKcyF6Qk)

This won't error because the `_slime.Width` property is accessed *after* the call to `base.Initialize()`.  When `base.Initialize()` is called, internally the last thing it does is call `LoadContent` before it returns back, so the slime sprite is already loaded before it's width property is accessed.  These initializations could have just been added inside the `LoadContent` method instead of splitting them between `LoadContent` and `Initialize`, but I choose to do it this way on purpose since most people aren't aware of when `LoadContent` is actually called in relation to the `base.Initialize()`   This is covered in [Content Loading](https://aristurtle.net/tutorials/building_2d_games/03_the_game1_file/#content-loading) section of chapter 03.

djjohnsongeek commented:
> I've updated the chapters to (hopefully) resolve the issues if you wanna double check again.

I did, looks good!

> This won't error because the `_slime.Width` property is accessed after the call to `base.Initialize()`. 

Ah okay, this was my problem, I had  `base.Initialize();` at the very end of my Initialize method. I should have read Chapter 03 closer.

@AristurtleDev  You have done a great job! I love how this tutorial not only teaches how to use MonoGame, but also how to structure your game project well. I'll keep working through the rest of the chapters and post anything else I notice.

I have learned quite a few things I did not know before.
- I did not know about `Vector2.Reflect()`
- Or about `HashCode`, that is very handy
- Also its the first time I have learned about the Finalizer method
- And clearly I did not understand that `Initialize()` calls `LoadContent()`

djjohnsongeek commented:
Chapter 17, [The Game Scene](https://aristurtle.net/tutorials/building_2d_games/17_scenes/#the-game-scene). The code given does not include the changes made to the update and draw logic from [Chapter 12: Working With Tile Maps](https://aristurtle.net/tutorials/building_2d_games/13_working_with_tilemaps/). The the map graphics are missing.
![image](https://github.com/user-attachments/assets/cbdcb69c-f5d4-4c2b-9dd2-551836be0feb)
![image](https://github.com/user-attachments/assets/760ef127-389e-4014-a83f-852f1e49a265)
![image](https://github.com/user-attachments/assets/e1ab7d04-d928-4ae9-91e3-ce488632af84)
Simply adding the Tile Map reveals that the screen bounds and score text positions are also off.
![image](https://github.com/user-attachments/assets/630c5582-9799-49bd-8e5d-9abd31743645)

djjohnsongeek commented:
Chapter 18, [Border Color](https://aristurtle.net/tutorials/building_2d_games/18_texture_sampling/#border-color).
The text description does not match the image, or the image description.
> if we use the checkerboard pattern again, using Border Color mode with a border color of red, then it would produce the following ...

However the following is displayed:
![image](https://github.com/user-attachments/assets/a174a86e-60ec-439d-98e4-a6d82221a26d)

djjohnsongeek commented:
Okay that is everything I have noticed going though all the chapters so far. Curious if a simple 2D Camera is going to be covered? Fill free to ping me when the next chapter comes out, I would be happy to take a look!

AristurtleDev committed:
Fix errors in chapter 17 from retroactivly adding chapters

AristurtleDev committed:
Red to green

AristurtleDev commented:
@djjohnsongeek Thanks again for checking.

Chapter 17 should be fixed now.  This is another one where I retroactively added in a back chapter and missed updating the chapters that would come after it.

As for the image in chapter 18, I've updated the text in the paragraph to read "border color of green".  I Initially wrote the text with the idea to use red, but when I made the actual images for it, I used the MonoGame orange for the checkerboard pattern, and the red against that didn't contrast enough for the example, so I switched it to green.

AristurtleDev commented:
Also forgot to mention @djjohnsongeek 

This tutorial series itself will not have a 2D camera.  Cameras are somewhat of a basic topic, but they do get into matrix maths and fall a little out of scope for what this series is ultimately trying to teach.

Once this is finished for the MonoGame foundation bounty, i do plan to add additional series for a more intermediate level and advanced level in 2D that would go after reading through this one.  The intermediate one will include a simple camera.  It will be based on a [past tutorial](https://github.com/manbeardgames/manbeardgames.github.io/blob/gh-pages-develop/docs/tutorials/2d-camera.md)  I wrote a long time ago that needs to be updated, so I'm going to use that as the opportunity to update it lol.

djjohnsongeek commented:
@AristurtleDev

> Chapter 17 should be fixed now. This is another one where I retroactively added in a back chapter and missed updating the chapters that would come after it. As for the image in chapter 18 ...

Sounds good!

> This tutorial series itself will not have a 2D camera. Cameras are somewhat of a basic topic, but they do get into matrix math and fall a little out of scope for what this series is ultimately trying to teach.

I have never delved into matrix math myself, so I end up using 2D cameras without understanding them fully. That is on me for not looking into it. I still think there is a good argument for including a basic 2D camera implementation that just hand waives the matrix explanation. But I also get your point, no worries.

Final thought, one issue with any tutorial is that it can be tempting to copy everything (even if you are just writing it out) and not fully understand what is actually happening. It might be worth it to add a small implementation challenge to certain chapters that helps students get their hands dirty. It would be completely up to the student how to implement each challenge.
Please feel free to reject this as an idea, I don't want to step on your toes at all. I just got excited about it ...

- **Chapter 03 - The Game1 File:** Change the screen color.
- **Chapter 05 - Content Pipeline:** Load a second image and place it in the bottom right of the screen.
- **Chapter 06 - Working With Textures:** Draw two different sized textures on top of each other and rotate them opposite directions. This might need a hint to user the `Update()` method.
- **Chapter 07 - Optimize Texture Rendering:** Add your own texture to the atlas and draw it on the screen.
- **Chapter 09 - The AnimatedSprite Class:** Add your own new frame to one of the animations.
- **Chapter 11 - Input Management:** Add a new ability that makes the slime teleport in the direction it is moving when a mouse button or key is pressed.
- **Chapter 12 - Collision Detection:** Add a obstacle to the middle of the screen, have the bat bounce off of it, and don't allow the slime to move through it.
- **Chapter 14 - Working With Tilemaps:** Add support for solid tiles.
- **Chapter 15 - Audio Controller:** Add a button to the screen that lets you mute the games audio.
- **Chapter 16 - Working With SpriteFonts:** Not sure ...
- **Chapter 17 - Scenes:** Add a new type of bat that you have avoid. If your slime gets hit, go to a game over screen.
- **Chapter 17 - Scenes:** Add multiple bats to the game. Hint, uses `List`!
- **Chapter 18 - Texture Sampling:** No idea here actually. This chapter was almost full of brand new information to me.

SimonDarksideJ commented:
I like your idea @djjohnsongeek, make the student sweat :D

SimonDarksideJ commented on articles/tutorials/building_2d_games/18_texture_sampling/index.md:
```diff
@@ -0,0 +1,229 @@
+description: "Learn how to use texture sampling states and add a scrolling background effect to the game."
+---
+
+In previous chapters, we've drawn individual sprites and textures with the sprite batch, but for creating repeating background patterns, we need a more efficient approach than manually drawing the same texture multiple times. We could reuse the tilemap system that was created to make repeated background patterns, but this has a limitation in that the tiles are stationary and would require constan ...
```
Include link to chapter where this was done

SimonDarksideJ replied:
In two places
* drawing sprites
* Drawing tilemap

AristurtleDev replied:
Added links

SimonDarksideJ commented on articles/tutorials/building_2d_games/18_texture_sampling/index.md:
```diff
@@ -0,0 +1,229 @@
+
+## Understanding Texture Sampling
+
+When a texture is drawn to the screen, MonoGame uses a process called "sampling" to determine which pixels from the texture should be displayed, especially when the texture is scaled, rotated, or only partially visible. The rules governing this process are defined by sampler states.
```
Link to definition or explaination?

AristurtleDev replied:
Sampling definition has been expanded

SimonDarksideJ commented on articles/tutorials/building_2d_games/18_texture_sampling/index.md:
```diff
@@ -0,0 +1,229 @@
+
+This normalized system means that regardless of whether your texture is 32×32 pixels or 2048×2048 pixels, the coordinates to access the entire texture always range from 0.0 to 1.0. The graphics hardware automatically converts these normalized coordinates to the actual pixel locations within the texture.
+
+When you use SpriteBatch to draw a texture with a source rectangle specified in pixels, MonoGame internally converts those pixel coordinates to normalized texture coordinates before sending them to the GPU. Similarly, when you specify a destination rectangle, MonoGame determines how the normalized texture coordinates should map to screen coordinates.
```
Link to SPriteBatch API docs

AristurtleDev replied:
xref added for SpriteBatch

SimonDarksideJ commented on articles/tutorials/building_2d_games/18_texture_sampling/index.md:
```diff
@@ -0,0 +1,229 @@
+
+### Filtering Modes
+
+One aspect of sampler states if the filtering mode.  The filter mode used determines how pixels are interpolated when a texture is scaled.  There are three filtering modes available; Point, Linear, and Anisotropic.
```
Do not assume the reader understands these terms :D
Either expand or include a link for where to learn more

AristurtleDev replied:
Filtering definition as been expanded.

SimonDarksideJ commented on articles/tutorials/building_2d_games/18_texture_sampling/index.md:
```diff
@@ -0,0 +1,229 @@
+### Filtering Modes
+
+One aspect of sampler states if the filtering mode.  The filter mode used determines how pixels are interpolated when a texture is scaled.  There are three filtering modes available; Point, Linear, and Anisotropic.
+
```
Also needs to point out the performance impacts of each filtering mode

AristurtleDev replied:
Added performance impact of each filtering mode.

SimonDarksideJ commented on articles/tutorials/building_2d_games/18_texture_sampling/index.md:
```diff
@@ -0,0 +1,229 @@
+
+### Addressing Modes
+
+The other aspect is the addressing mode, which determines what happens when texture coordinates fall outside the normal 0.0 to 1.0 range.  There are four addressing modes available; Wrap, Mirror, Clamp, and Border Color:
```
Again, explain more about the term or link to a resource to explain

AristurtleDev replied:
Expanded explanation of adressing

SimonDarksideJ commented on articles/tutorials/building_2d_games/18_texture_sampling/index.md:
```diff
@@ -0,0 +1,229 @@
+
+The other aspect is the addressing mode, which determines what happens when texture coordinates fall outside the normal 0.0 to 1.0 range.  There are four addressing modes available; Wrap, Mirror, Clamp, and Border Color:
+
+#### Wrap Mode
```
This section by far shows we chose the right person for the job for this tutorial, great work explaining all the options (and not doing "just do as I say")

AristurtleDev replied:
Thanks :D

SimonDarksideJ commented on articles/tutorials/building_2d_games/18_texture_sampling/index.md:
```diff
@@ -0,0 +1,229 @@
+- The `_backgroundPattern` field was added to store a reference to the texture of the background pattern once its loaded.
+- The `_backgroundDestination` field was added to define the destination rectangle to draw the background pattern to.
+- The `_backgroundOffset` field was added to apply an offset to the source rectangle when rendering the background pattern to give it the appearance that it is scrolling.
+- The `_scrollSpeed` field was dded to set the speed at which the background pattern scrolls.
```
typo "added"

AristurtleDev replied:
Typo fixed.

SimonDarksideJ reviewed (changes requested):
A few comments

AristurtleDev commented:
> Final thought, one issue with any tutorial is that it can be tempting to copy everything (even if you are just writing it out) and not fully understand what is actually happening. It might be worth it to add a small implementation challenge to certain chapters that helps students get their hands dirty. It would be completely up to the student how to implement each challenge. Please feel free to reject this as an idea, I don't want to step on your toes at all. I just got excited about it ...

This is definitely something I can go back and add in to each chapter.  It'll most likely be once everything is at the final stages so i can focus on finishing up the remaining chapters first, but I like this idea.

AristurtleDev committed:
Include chapter links

AristurtleDev committed:
Add xref to SpriteBatch

AristurtleDev committed:
Correct spelling of added

AristurtleDev committed:
Explain more in detail what sampling is

AristurtleDev committed:
Explain i more detail what filtering is.

AristurtleDev committed:
Add performance considerations for each filter mode.

AristurtleDev committed:
Explain more in detail what addressing is.

AristurtleDev commented:
I've updated the feedback provided from @SimonDarksideJ, going to go ahead and merge and ask for forgiveness later so I can start the PR for chapter 19.
